ingest/src/main/java/com/microsoft/azure/kusto/ingest/resources/RankedStorageAccount.java

```diff
@@ -53,16 +61,21 @@ private Bucket adjustForTimePassed() {
             return buckets.peek();
         }
 
-        if (bucketsToCreate >= maxNumberOfBuckets) {
-            buckets.clear();
-            buckets.push(new Bucket());
-        } else {
-            for (int i = 0; i < bucketsToCreate; i++) {
+        try {
+            bucketsLock.writeLock().lock();
```

Comment:
Isn't there a better construct for locking/unlocking (try-with-resources?) or does the lock not support it

Reply:
If you want we can create a closeable lock ourselves

Reply:
Would an AtomicReference work instead? Thread-safe but no lock needed.

Reply:
We could use `synchronized` blocks which look a bit simpler, but the benefit of this is that read locks don't prevent other threads from reading.

Regarding `AtomicReference`, thanks for your suggestion, @The-Funk! But I'm not sure the code would be simpler in this scenario because we're adding thread safety to an `ArrayDeque`, which is a collection that isn't itself thread-safe (which we might be able to change, but I don't know if the code would be simpler).